### DISCUSSION TITLE
How to Disable Transparent Huge Pages in Github Actions

### DISCUSSION THREAD

**Discussion by mckalexee** [community/community#26004] [Actions]
<p>I’m trying to run some PHP tests for a legacy application, and part of that process requires running a container for MariaDB that uses TokuDB (very obscure engine)</p>
<p>The issue is that TokuDB doesn’t work if Transparent Huge Pages are enabled.</p>
<p>How can I disable THP for Github Actions? I’m wanting to move away from Circle CI but this is the last thing blocking me.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/mckalexee">@mckalexee</a>,</p>
<p>To check if Transparent Hugepages is enabled on the Linux machine, you can run the following command.</p>
<pre><code class="lang-auto">cat /sys/kernel/mm/redhat_transparent_hugepage/enabled
</code></pre>
<ul>
<li>If it is enabled, the returned setting is this.<pre><code class="lang-auto">[always] madvise never
</code></pre>
</li>
<li>If it is disabled, the returned setting is this.<pre><code class="lang-auto">always madvise [never]
</code></pre>
</li>
</ul>
<p>To disable Transparent Hugepages on the Linux runner machine:</p>
<ol>
<li>
<p>If on the GitHub-hosted runner, you can try to re-install the Hugepages and use the option “<strong>–thp-never</strong>” to disable Transparent Hugepages.</p>
<pre><code class="lang-yaml">- name: Check Transparent Hugepages
  run: cat /sys/kernel/mm/transparent_hugepage/enabled

- name: Re-install Hugepages and disable Transparent Hugepages
  run: |
    sudo su
    sudo apt-get install hugepages
    sudo hugeadm --thp-never

- name: Check Transparent Hugepages again
  run: cat /sys/kernel/mm/transparent_hugepage/enabled
</code></pre>
</li>
<li>
<p>If on your self-hosted runner, generally you can follow one one of the two methods below:</p>
<ul>
<li>Edit your boot loader (such as <strong>/etc/grub.conf</strong>) and add the following line to the end of the kernel line.<pre><code class="lang-auto">transparent_hugepage=never
</code></pre>
</li>
<li>Edit <strong>/etc/rc.local</strong> and add the following script.<pre><code class="lang-auto">if test -f /sys/kernel/mm/redhat_transparent_hugepage/enabled; then
  echo never &gt; /sys/kernel/mm/redhat_transparent_hugepage/enabled
fi
</code></pre>
</li>
</ul>
<p>After one of above two methods, you need to reboot your system for the setting to take effect, or run the following echo line to proceed with the install without rebooting:</p>
<pre><code class="lang-auto">echo never &gt; /sys/kernel/mm/redhat_transparent_hugepage/enabled
</code></pre>
</li>
</ol>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/mckalexee">@mckalexee</a>,</p>
<p>How are things going?<br>
Have you tried my above suggestions to disable Transparent Hugepages on the Linux runner machine?<br>
Are they helpful to you?<br>
Please have try, any update, feel free to tell us.</p>

**Comment by mckalexee** [comment]
<aside class="quote group-github-partners" data-username="brightran" data-post="2" data-topic="127769">
<div class="title">
<div class="quote-controls"></div>
 brightran:</div>
<blockquote>
<pre><code class="lang-auto">  sudo su
    sudo apt-get install hugepages
    sudo hugeadm --thp-never
</code></pre>
</blockquote>
</aside>
<p>Unfortunately this doesn’t work on the GitHub-hosted runner. The container has to be running in privileged mode to modify THP.</p>

**Comment by airtower-luna** [comment]
<p>It should work directly in the runner VM, but to ensure the change is made before the container is started you’ll have to manage your container(s) directly instead of relying on job-level <code>container</code> or <code>services</code> settings.</p>

**Comment by mckalexee** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a> <a class="mention" href="/brightran">@brightran</a></p>
<p>That was it. Thanks!</p>
<p>I needed to manually control the containers instead of using the services option.</p>
<p>One small pain point is that I can’t run my tasks in my own docker container. This means I have to install some things in the linux runner every time, and the actions testing environment doesn’t 100% match my production environment any more.</p>
<p>I can work around it just fine, but it would be nice in the future if a flag existed to disable THP out the gate so I can use some of the built in docker features.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="mckalexee" data-post="6" data-topic="127769">
<div class="title">
<div class="quote-controls"></div>
 mckalexee:</div>
<blockquote>
<p>One small pain point is that I can’t run my tasks in my own docker container. This means I have to install some things in the linux runner every time, and the actions testing environment doesn’t 100% match my production environment any more.</p>
</blockquote>
</aside>
<p>Well, you could turn the current steps into a script running in yet another Docker container. Though admittedly that loses the convenience of including pre-made actions.</p>
